### ISSUE TITLE
Method for confirming clusterrepos are created and downloaded in Rancher v2.5 clusters

### ISSUE THREAD

**Issue by axeal** [rancher/terraform-provider-rancher2#590]
Attempting to deploy a chart from the `rancher-charts` or `rancher-partner-charts` repositories in a Rancher v2.5 provisioned cluster, via the `rancher2_app_v2` resource, immediately after cluster provisioning will fail with a message of the following format, if the clusterrepos has not yet been created.

```
Error: Bad response statusCode [404]. Status [404 Not Found]. Body: [code=NotFound, message=clusterrepos.catalog.cattle.io "rancher-charts" not found] from [https://178.128.253.220/k8s/clusters/c-wmhlg/v1/catalog.cattle.io.clusterrepos/rancher-charts] 
```

User's request is to have a method of verifying that the clusterrepo has been created and downloaded, so that the creation of the `rancher2_app_v2` is only attempted when this is true. One workaround is to use the `state_confirm` argument of the `rancher2_cluster_sync` resource, but this is arbitrary and cannot guarantee the repository is available yet.

gz#15112

**Comment by rawmind0** [comment]
Added retry logic to get V2 catalogs and apps at PR https://github.com/rancher/terraform-provider-rancher2/pull/569 . This should address the issue.

**Comment by rawmind0** [comment]
Fix included on new provider release v1.12.0

Please, reopen issue if needed

**Comment by julienym** [comment]
This is still a problem with 1.12.0, but in our case it's not happening all the time.

Error: Bad response statusCode [404]. Status [404 Not Found]. Body: [code=NotFound, message=clusterrepos.catalog.cattle.io "rancher-charts" not found] from [https://XXXXXXXXXXXXXXX/k8s/clusters/c-57c5v/v1/catalog.cattle.io.clusterrepos/rancher-charts]

Why not use a data source and verify a refreshed attribute?

**Comment by julienym** [comment]
@rawmind0 could you reopen this please?

**Comment by rawmind0** [comment]
@julienym there is no `refreshed` attribute for clusterrepos, just a [`Downloaded` condition](https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_catalog_v2.go#L147). This condition is checked when you add a new clusterrepo, waiting until the clusterrepo is downloaded https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_catalog_v2.go#L45

If it's not happening always is due to a race condition. Latest patch added few retries (3 retries sleeping 5 seconds) on appV2 create, seems that clusterrepos download is lasting more than 15 seconds eventually.  Could you please explain a little bit more about your use case?? Are you using  `cluster_sync` resource??

**Comment by julienym** [comment]
Hi @rawmind0 
In our case, we can't add the repos since it's a default one coming with Rancher 2.5 - > rancher-charts

We do use cluster_sync, here is the resource:
```
resource "rancher2_app_v2" "rancher_gatekeeper_operator" {
  count = var.enable_opa_client ? 1 : 0

  cluster_id    = rancher2_cluster_sync.this.id
  name          = "rancher-gatekeeper"
  namespace     = rancher2_namespace.gatekeeper-system[count.index].id
  repo_name     = "rancher-charts"
  chart_name    = "rancher-gatekeeper"
  chart_version = var.gatekeeper_chart_version
}
```

**Comment by rawmind0** [comment]
Got it, the issue is with default clusterrepos. thanks for the info @julienym 

Working on a patch

**Comment by rawmind0** [comment]
Submitted PR https://github.com/rancher/terraform-provider-rancher2/pull/609 to fix the issue

**Comment by rawmind0** [comment]
Fix included at new Rancher2 tfp release v1.13.0
